### PR TITLE
GraphEngine python binding - core components

### PR DESCRIPTION
**Pull request by thautwarm**
dynamic loading completed

### CODE REVIEW

**Comment by yatli** [comment]
reviewing

File: src/Modules/DynamicLoading/DynamicLoading/Center.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                    string asmLoadDir = null,
+                    string versionName = null)
+        {
+
+            CurrentVersion = new AsmVersion(
+                _currentCellTypeOffset,
+                versionName ?? DateTime.Now.ToString(),
+                tslsrcDir, tslBuildDir?? tslsrcDir, asmLoadDir??tslBuildDir??tslsrcDir);
+
+            Console.WriteLine("tslsrcDir: " + tslsrcDir +
```
Comment (yatli) [inline src/Modules/DynamicLoading/DynamicLoading/Center.cs]: refactor with Log.WriteLine

File: GraphEngine/PluginConfig.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+﻿using System;
```
Comment (yatli) [inline GraphEngine/PluginConfig.cs]: Why is this project moved?

File: src/Modules/DynamicLoading/DynamicLoading/Center.cs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            public static void CmdCall(string cmd, string arguments)
+            {
+                Console.WriteLine("command:  " + cmd + " " + arguments);
+                Process proc = new Process();
+                proc.StartInfo.FileName = cmd;
+                proc.StartInfo.Arguments = arguments;
+                proc.StartInfo.UseShellExecute = false;
+                proc.StartInfo.RedirectStandardOutput = true;
+                proc.Start();
+                Console.WriteLine(proc.StandardOutput.ReadToEnd());
```
Comment (yatli) [inline src/Modules/DynamicLoading/DynamicLoading/Center.cs]: refactor with Log.WriteLine

File: src/Modules/DynamicLoading/DynamicLoading/Center.cs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                        CurrentVersion.TslBuildDir,
+                        $"{Namespace}.csproj");
+            using (var s = System.IO.File.CreateText(path))
+            {
+                CSProj.TemplateLine.ForEach(s.WriteLine);
+            }
+        }
+        public static bool CodeGenFn()
+        {
+            System.IO.Directory.GetFiles(CurrentVersion.TslSrcDir, "*.tsl").ToList().ForEach(Console.WriteLine);
```
Comment (yatli) [inline src/Modules/DynamicLoading/DynamicLoading/Center.cs]: same here

File: src/Modules/DynamicLoading/DynamicLoading/Center.cs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+            if (!BuildFn())
+            {
+                throw new TSLBuildError();
+            }
+
+            try
+            {
+                var asm = LoadFn();
+                asm.GetTypes().ToList().ForEach(Console.WriteLine);
```
Comment (yatli) [inline src/Modules/DynamicLoading/DynamicLoading/Center.cs]: for point-free log.writeline, you can wrap up your own logging function within the class, like `void _log(string content) => Log.WriteLine(LogLevel.Info, $"{nameof(your_class)}: {0}", content);`

File: src/Modules/DynamicLoading/DynamicLoading/Center.cs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                _asmSlides.Add(asm);
+
+                /// TODO: get cell_type count by `asm`.
```
Comment (yatli) [inline src/Modules/DynamicLoading/DynamicLoading/Center.cs]: use AssemblyUtility.GetAllClassInstances<IStorageSchema> to extract the schema from the generated assembly.

File: src/Modules/Trinity.FFI/Trinity.FFI.Native/Trinity.FFI.SWIG.h [hunk 7]
```diff
@@ -42,6 +42,16 @@ class Cell
...
+
+	void SetID() {
+		g_TrinityInterfaces->cell_setid(m_cell, GetID());
+	}
+
+
+
 	~Cell()
 	{
 		printf("hey dtor\n");
```
Comment (yatli) [inline src/Modules/Trinity.FFI/Trinity.FFI.Native/Trinity.FFI.SWIG.h:57]: dd

File: src/Modules/Trinity.FFI/Trinity.FFI.Native/Trinity.FFI.SWIG.h [hunk 8]
```diff
@@ -42,6 +42,16 @@ class Cell
 		return g_TrinityInterfaces->cell_hasfield(m_cell, field);
 	}
 
+	long long GetID() {
+		return g_TrinityInterfaces->cell_getid(m_cell);
+	}
+
+	void SetID() {
+		g_TrinityInterfaces->cell_setid(m_cell, GetID());
```
Comment (yatli) [inline src/Modules/Trinity.FFI/Trinity.FFI.Native/Trinity.FFI.SWIG.h:50]: SetID with no incoming cellid?

File: src/Modules/Trinity.FFI/Trinity.FFI.PythonModule/collect.py [hunk 9]
```diff
@@ -1,60 +1,83 @@
 """
```
Comment (yatli) [inline src/Modules/Trinity.FFI/Trinity.FFI.PythonModule/collect.py:1]: 👍

File: src/Modules/Trinity.FFI/Trinity.FFI.PythonModule/collect.py [hunk 10]
```diff
@@ -1,60 +1,83 @@
...
+                    to_module='ffi')
+
+    # copy exe
+    to_pymodule_dir(pardir_of(CORECLR_PATH),
+                    with_postfix=['.exe'],
+                    to_module='Command')
+
+    # add newtonsoft.json.dll
+    # TODO: a robust way to find out newtonsoft.json.dll.
+    to_pymodule_dir(r'C:\Program Files\dotnet\sdk\NuGetFallbackFolder\newtonsoft.json\9.0.1\lib\netstandard1.0',
```
Comment (yatli) [inline src/Modules/Trinity.FFI/Trinity.FFI.PythonModule/collect.py]: if we assume 'dotnet' to be an available command, we can use it to query the location of the package cache.

**Review by yatli — CHANGES_REQUESTED** [review]
good job red! let's work together to fix the rough edges. I'm adding some bits into the codegen as we discussed yesterday.

**Comment by thautwarm** [comment]
I thought an interface method `LoadGenericCell ` for `IGenericCellOperations` with arguments (storage, cellId, buff) is badly needed.
`Use, New, Save ` all load from storage only once, however there is not a proper method to apply the same way to `Save`.
To detail, we need
```C#
public unsafe ICell LoadGenericCell(byte[] buff, long cellId, ICell cell)
{
    switch ((CellType)type)
    {
        case CellType.xxx:
                fixed (byte* xxx_ptr = buff)
                    {
                        xxx_Accessor xxx_accessor = new xxx_Accessor(xxx_ptr);
                        xxx_accessor.CellID = cellId;
                        return (xxx)xxx_accessor;
                    }
               ...
    }
}
```

File: src/Modules/Trinity.FFI/Trinity.FFI.Native/Trinity.FFI.SWIG.h (hunk 7 above)
Comment (thautwarm) [reply src/Modules/Trinity.FFI/Trinity.FFI.Native/Trinity.FFI.SWIG.h:57]: qaq

**Review by TaviTruman — COMMENTED** [review]
Great design facet correction and overall enhancement. In general server-side services should at least be asynchronous if not multitasking.

**Comment by yatli** [comment]
well done!

**Pull request merged**